OA-PASS/pass-docker
# User service and shibboleth with real data

# Overview
* Uses a shared volume between Fedora and elasticsearch that allows mounting pre-loaded assets;  Fedora and elasticsearch now contain the full bootstrapped PASS dataset
* Configures shibboleth so that the proper attributes are released to Fedora and the User Service.  The user service is available at `https://pass/pass-user-service/whoami`
* Changes the set of users that can log in via shibboleth (`https/pass/*`).  All users have the same password `moo`
  * `staff1` A staff member who _does_ have grants and therefore _does_ have a `User` resource already, and is the PI on grants.
  * `staff2` A staff member who _does not_ have grants.  Because PASS policy is to only allow _faculty_ in, this user will be _denied_ access to the user service, or the repository.
  * `faculty1` A faculty member who _does_ have lots of grants, and therefore _does_ have a `User` resource already, and is a PI on grants.
  * `faculty2` A faculty member who _does not_ have grants.  Because PASS policy is to give faculty submitter privileges, the first time this person logs to shibboleth and hits the user service, a new `User` resource is created, but the user _should not_ be allowed to associate grants with his or her submissions.  

# To test: 
* Check out this PR
* Make sure you are given access to `birkland/assets`, as this is hosted privately.  Give Aaron your dockerhub username (or create one)
* Make sure you have followed the [prerequisite](https://github.com/OA-PASS/pass-docker#prereq) instructions and added `pass` to your hosts file, so that the address resolves
* do a `docker-compose pull`
* `docker-compose up`
* Once everything has settled down, go to the user service, `https://pass/pass-user-service/whoami`
  * you will be redirected to shibboleth login.  Pick one of the users and log in.  You'll be redirected back to the user service, and you should see sensible JSON
  * If you want, follow the link to the `User` resource in Fedora (the `@id` in the JSON)

# Known Issues
* As usual in `pass-docker`, only users defined in [tomcat-users.xml](https://github.com/birkland/pass-demo-docker/blob/bootstrap-assets/fcrepo/4.7.5-SNAPSHOT/conf/tomcat-users.xml) can log in.  These are the users listed above.  Currently, real JHU shibboleth is also restricted to these users, so that needs to be fixed in a later PR.
* Sometimes the shibboleth idP may fail to start due to #25.  It's probably time to look at that.  Until then, just note that on random occasions (particularly after a `docker stop` and re-start) shib might not work after startup. 
* Ember uses `findAll`, which causes severe resource consumption by all parties and usually causes errors.  See OA-PASS/pass-ember#348
* It takes a long time for Ember to get up and running due to its deposit of test data.  This deposit should probably be disabled

# Other considerations
This PR places Fedora and Elasticsearch data into a _volume_.  This has some consequences
  * `docker-compose down` will destroy the containers _but keep the volume intact_.  So any modifications to the data will be persistent, even if the containers are not.  This differs significantly from the old behaviour, when `docker-compose down` would wipe everything clean (since state was kept in containers)
* To totally blow away all changes to the data and start fresh, do `docker-compose down -v`, or `docker volume prune`.  The next time `docker-compose` is started, it will use the stock assets as pulled from the private repository.
